### PR TITLE
Fixed builds for specific runners

### PR DESCRIPTION
**Pull request by ayufan**
Don't run projects which are assigned to specific runner by shared runner.

### CODE REVIEW

**Comment by ayufan** [comment]
@randx @vsizov any update?

File: app/views/admin/runners/index.html.haml [hunk 1]
```diff
@@ -23,10 +23,10 @@
     %ul
       %li
         %span.label.label-success shared
-        \- run any builds
+        \- run builds from all non-assigned projects
```
Comment (dblessing) [inline app/views/admin/runners/index.html.haml]: Non-assigned -> unassigned

File: app/views/admin/runners/show.html.haml [hunk 2]
```diff
@@ -9,12 +9,12 @@
 
 - if @runner.shared?
   .bs-callout.bs-callout-success
-    %h4 This runner will process build from ALL projects
+    %h4 This runner will process build from ALL NON ASSIGNED projects
```
Comment (dblessing) [inline app/views/admin/runners/show.html.haml]: Non-assigned -> unassigned in this file, too.

File: app/views/projects/_no_shared_runners.html.haml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+.alert.alert-danger
+  %p
+    There is NO runners to build this project. Consider assigning this project to
```
Comment (dblessing) [inline app/views/projects/_no_shared_runners.html.haml]: There are NO runners

File: app/views/projects/_no_shared_runners.html.haml [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+    %span.label.label-info specific
```
Comment (dblessing) [inline app/views/projects/_no_shared_runners.html.haml]: a specific

File: app/views/projects/_specific_runner.html.haml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+.alert.alert-info
+  %p
+    This project uses
+    %span.label.label-info specific
```
Comment (dblessing) [inline app/views/projects/_specific_runner.html.haml]: a specific

**Comment by dblessing** [comment]
@ayufan This looks nice. I made a few comments on grammar.

**Comment by ayufan** [comment]
@dblessing Thanks. Fixed.

File: app/views/admin/runners/show.html.haml [hunk 6]
```diff
@@ -9,12 +9,12 @@
...
 - if @runner.shared?
   .bs-callout.bs-callout-success
-    %h4 This runner will process build from ALL projects
+    %h4 This runner will process build from ALL UNASSIGNED projects
     %p If you want runner to build only specific projects you just need to enable them in table below
 - else
   .bs-callout.bs-callout-info
     %h4 This runner will process build only from ASSIGNED projects
-    %p If you want runner to build all projects you just need to disable all assignee projects
+    %p If you want runner to build non assigned projects you just need to disable all assignee projects
```
Comment (dblessing) [inline app/views/admin/runners/show.html.haml]: Whoops, missed this `non-assigned`.

File: app/views/projects/_no_shared_runners.html.haml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+.alert.alert-danger
+  %p
+    There are NO runners to build this project. Consider assigning this project to
+    %span.label.label-info a specific
```
Comment (dblessing) [inline app/views/projects/_no_shared_runners.html.haml]: What does this look like in the UI? Does it make sense to have `a specific` inside the label or on the line before? Same with the specific runner partial.

Comment (ayufan) [inline app/views/projects/_no_shared_runners.html.haml]: You're right. It doesn't make sense. I just wanted to have the same label
as in runners admin view.


On Fri, Jan 2, 2015 at 3:59 PM, Drew Blessing <notifications@github.com>
wrote:

> In app/views/projects/_no_shared_runners.html.haml
> <https://github.com/gitlabhq/gitlab-ci/pull/488#discussion-diff-22411917>:
>
> > @@ -0,0 +1,5 @@
> > +.alert.alert-danger
> > +  %p
> > +    There are NO runners to build this project. Consider assigning this project to
> > +    %span.label.label-info a specific
>
> What does this look like in the UI? Does it make sense to have a specific
> inside the label or on the line before? Same with the specific runner
> partial.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gitlabhq/gitlab-ci/pull/488/files#r22411917>.
>



-- 
Kamil Trzciński

ayufan@ayufan.eu
www.ayufan.eu

**Comment by ayufan** [comment]
Fixed once again :)

**Comment by dblessing** [comment]
Cool, great work. Thanks.

**Comment by randx** [comment]
@vsizov can you review this please?

File: app/helpers/projects_helper.rb [hunk 8]
```diff
@@ -22,4 +22,16 @@ def html_badge_code(project, ref)
     url = status_project_url(project, ref: ref, format: 'png')
     "<a href='#{project_url(project, ref: ref)}'><img src='#{url}' /></a>"
   end
+
+  def runners_for_project(project)
+    project.runners.each.map { |r| "#" + r.id.to_s }.join(", ")
```
Comment (vsizov) [inline app/helpers/projects_helper.rb]: let's remove `each` from here

File: app/helpers/projects_helper.rb [hunk 9]
```diff
@@ -22,4 +22,16 @@ def html_badge_code(project, ref)
...
+  end
+
+  def project_uses_specific_runner?(project)
+    project.runners.present?
```
Comment (vsizov) [inline app/helpers/projects_helper.rb]: any? is more suitable here

File: app/helpers/projects_helper.rb [hunk 10]
```diff
@@ -22,4 +22,16 @@ def html_badge_code(project, ref)
...
+  end
+
+  def no_shared_runners_for_project?(project)
+    Runner.count.nonzero? && project.runners.blank? && Runner.shared.count.zero?
```
Comment (vsizov) [inline app/helpers/projects_helper.rb]: why not just like
`Runner.shared.blank?` ?

**Comment by ayufan** [comment]
@vsizov updated

**Comment by ayufan** [comment]
@vsizov ?

**Pull request merged**